arch: xtensa: Print backtrace from panic handler

shubhamkulkarni97 opened the pull request:
This change uses stack frame to print backtrace once exception occurs.
Backtrace unwinding is supported for ESP32 as of now, but can be easily supported by other xtensa architectures.

This PR depends on https://github.com/zephyrproject-rtos/zephyr/pull/30424 and contains commits from the same.

shubhamkulkarni97 commented:
@dcpleung Can you please review top commit of this PR? I have kept this PR as a draft, as it includes commits from baseline PR.

dcpleung commented on arch/xtensa/include/xtensa-asm2-context.h:
```diff
@@ -101,11 +103,16 @@
+#define BSA_EXCCAUSE_OFF \
+			(BASE_SAVE_AREA_SIZE - \
+			 (BASE_SAVE_AREA_SIZE_COMMON + \
+			  BASE_SAVE_AREA_SIZE_LOOPS + 4))
```
What is `4` for? Is it `BASE_SAVE_AREA_SIZE_EXCCAUSE`?

shubhamkulkarni97 replied:
Yes, it is for `BASE_SAVE_AREA_SIZE_EXCCAUSE`. Have replaced it.

dcpleung commented on arch/xtensa/include/xtensa-asm2-context.h:
```diff
@@ -88,7 +89,8 @@
 	 BASE_SAVE_AREA_SIZE_LOOPS + \
 	 BASE_SAVE_AREA_SIZE_SCOMPARE + \
-	 BASE_SAVE_AREA_SIZE_THREADPTR)
+	 BASE_SAVE_AREA_SIZE_THREADPTR + \
```
It seems like you are inserting `EXCCAUSE` before `SCOMPARE`. Please keep this list ordered so it is easier to tell the ordering of these fields.

dcpleung commented on arch/xtensa/include/xtensa-asm2-context.h:
```diff
@@ -101,11 +103,16 @@
 			(BASE_SAVE_AREA_SIZE - \
 			 (BASE_SAVE_AREA_SIZE_COMMON + \
 			  BASE_SAVE_AREA_SIZE_LOOPS + \
+			  BASE_SAVE_AREA_SIZE_EXCCAUSE + \
```
You will also need to add this below inside `BSA_THREADPTR_OFF`.

dcpleung commented on include/arch/log.h:
```diff
@@ -0,0 +1,11 @@
+/* log.h - automatically selects the correct log.h file to include */
```
These doesn't seem to be a need for this file. You can simply include `arch/xtensa/log.h` directly as what you have is only for Xtensa and none of the other architectures.

dcpleung replied:
There really is not need for this file at this point, as what you have here is only for Xtensa (and not for multiple arch).

shubhamkulkarni97 replied:
Have removed this file as per suggestion

dcpleung commented on include/arch/xtensa/log.h:
```diff
@@ -0,0 +1,15 @@
+/*
```
Are all these just for printing exception messages? Or do you intend these to be used for all log messages? Because you define these in `include/arch/xtensa/log.h` which means these should be used for all logging in Xtensa.

shubhamkulkarni97 replied:
These changes are for all log messages in `xtensa-asm2.c` and `fatal.c`. These changes are required to handle [Cache disabled but cache memory region accessed](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-guides/fatal-errors.html#cache-disabled-but-cached-memory-region-accessed) exception.

Using LOG module in Zephyr will lead to flash cache access and cause another exception from panic handler.

dcpleung replied:
The problem with overriding `LOG_*()` is that the formatted output would not be consistent, since all the formatting code from the logging subsystem is bypassed. Also, by overriding `LOG_*()`, you no longer have the option to send the log message through, for example, the network logging backend. If remote logging is required, the exception messages (fatal and non-fatal ones) are no longer being sent remotely through some backends.

shubhamkulkarni97 replied:
* While flash operations like write, erase, etc. are performed on ESP32, flash cache and hence XIP is disabled. So, the code present in flash cache cannot be accessed. When a crash occurs in such scenario, code from flash cannot be accessed. That's why entire panic handler and all the dependent functions must be placed in internal RAM.
* LOG module is one of the dependencies of panic handler. Also, LOG module has a lot of dependencies. Hence, it is not possible to move it to internal RAM as it will eat up memory required for end application.
* In ESP-IDF, minimalistic printf implementation is used in panic handler which directly calls UART APIs and is redirected to ROM code. This is preferred because normal printf implementation can allocate dynamic memory and it is not recommended to allocate memory when system is crashed.
* As ESP32 support is in baseline mode currently, it will be very difficult to debug potential exceptions without these changes. Also, we are providing a configurable option to enable/disable backtrace feature. This config option can be enabled in debugging scenarios only.

@dcpleung Please let me know your views on this.

andrewboie replied:
please don't hack the definitions of LOG like this.

shubhamkulkarni97 replied:
Have moved log module to internal memory which fixes the issue. This change is removed.

dcpleung commented on arch/xtensa/core/xtensa_backtrace.c:
```diff
@@ -0,0 +1,120 @@
+	if (cause != EXCCAUSE_INSTR_PROHIBITED) {
+		mask = stk_frame.pc & 0xc0000000;
+	}
+	/* esp_cpu_get_backtrace_start(&stk_frame); */
```
remove?

dcpleung reviewed (changes requested):
Please separate the linker changes into another commit. So you will have one commit changing the Xtensa arch code, and another one for the ESP32 linker changes.

shubhamkulkarni97 committed:
arch: xtensa: Print backtrace from panic handler

This change uses stack frame to print backtrace once exception occurs
Printing backtrace helps to identify the cause of exception

Signed-off-by: Shubham Kulkarni <shubham.kulkarni@espressif.com>

shubhamkulkarni97 committed:
soc: esp32: linker: Fix backtrace for Invalid cache access exception

This change moves .rodata for panic handler and fatal.c into DRAM
Moves panic handler and its dependent functions into IRAM

Signed-off-by: Shubham Kulkarni <shubham.kulkarni@espressif.com>

shubhamkulkarni97 committed:
soc/esp32: Move logging library to IRAM with CONFIG_LOG_MINIMAL

This commit updates linker script to move minimal log module to internal
RAM

Signed-off-by: Shubham Kulkarni <shubham.kulkarni@espressif.com>

shubhamkulkarni97 committed:
soc/esp32: Move full logging library to IRAM

This commit updates linker script to move full logging module into
internal RAM

Signed-off-by: Shubham Kulkarni <shubham.kulkarni@espressif.com>

dcpleung commented on boards/xtensa/esp32/esp32_defconfig:
```diff
@@ -32,3 +32,5 @@ CONFIG_I2C_1=y
 CONFIG_CLOCK_CONTROL=y
 
 CONFIG_BOOTLOADER_ESP_IDF=y
+CONFIG_LOG=y
+CONFIG_LOG_DETECT_MISSED_STRDUP=n
```
I don't think these should be turned on by default.

shubhamkulkarni97 replied:
Removed as per suggestion

dcpleung commented:
Please fix the issues reported by the compliance checker.

shubhamkulkarni97 commented on soc/xtensa/esp32/linker.ld:
```diff
@@ -54,6 +54,9 @@ PROVIDE ( esp32_rom_g_ticks_per_us_app = g_ticks_per_us_app );
 PROVIDE ( esp32_rom_ets_delay_us = ets_delay_us );
 
+/* __udivdi3 is exported using assignment, which declares strong symbols */
+__udivdi3 = 0x4000cff8;
```
All the libgcc functions are provided in ROM. Here, `__udivdi3` is assigned to its ROM address and hence is true for all applications.

For reference: https://github.com/espressif/esp-idf/blob/8cd16b60f56dec4f319f6d2e73b6dec794588312/components/esp_rom/esp32/ld/esp32.rom.libgcc.ld#L87
